perf_benchmark/utils.cc

```diff
@@ -167,6 +167,77 @@ std::string GetStreamedJson(absl::string_view json_msg, uint64_t stream_size) {
   ss << ']';
   return ss.str();
 }
+
+// Copied from "test/test_common.h".
+std::string SizeToDelimiter(unsigned size) {
+  unsigned char delimiter[5];
+  // Byte 0 is the compression bit - set to 0 (no compression)
+  delimiter[0] = 0;
+  // Bytes 1-4 are big-endian 32-bit message size
+  delimiter[4] = 0xFF & size;
+  size >>= 8;
+  delimiter[3] = 0xFF & size;
+  size >>= 8;
+  delimiter[2] = 0xFF & size;
+  size >>= 8;
+  delimiter[1] = 0xFF & size;
+
+  return std::string(reinterpret_cast<const char*>(delimiter),
+                     sizeof(delimiter));
+}
+
+std::string WrapGrpcMessageWithDelimiter(absl::string_view proto_binary) {
+  return absl::StrCat(SizeToDelimiter(proto_binary.size()), proto_binary);
+}
+
+// The caller will need to take the ownership of the returned object by properly
+// releasing the memory.
+NestedPayload* GetNestedPayloadRecursive(uint64_t layers,
```

Comment:
You should return unique_ptr.

auto proto = std::make_unique<NestedPayload>();

Reply:
I couldn't do that because this function is generating NestedPayload recursively. `proto->set_allocated_nested()` only takes in a raw pointer. If I wrap it with `unique_ptr` then the pointer would be dangling after the recursive call returned.

Please see my other comment https://github.com/grpc-ecosystem/grpc-httpjson-transcoding/pull/80#discussion_r974506653.

Reply:
This `GetNestedPayloadRecursive` isn't exposed in the header file. `GetNestedPayload` always returns a unique_ptr, so it should be safe.

Reply:
I guess it's safe, but still is harder to understand ownership. 

If you return unique_ptr, this change should work:

```
proto->set_allocated_nested(GetNestedPayloadRecursive(...).release());
```

https://en.cppreference.com/w/cpp/memory/unique_ptr/release

Reply:
Thanks! I didn't know unique_ptr has this method. Just gained some more C++ knowledge today :)